PreMailer.Net/PreMailer.Net/CssElementStyleResolver.cs

```diff
@@ -8,20 +8,44 @@ public static class CssElementStyleResolver
 	{
 		private const string premailerAttributePrefix = "-premailer-";
 
-		public static IEnumerable<AttributeToCss> GetAllStyles(IElement domElement, StyleClass styleClass)
-		{
-			var attributeCssList = new List<AttributeToCss>();
+	public static IEnumerable<AttributeToCss> GetAllStyles(IElement domElement, StyleClass styleClass)
+	{
+		var attributeCssList = new List<AttributeToCss>();
+		var originalStyleAttr = domElement.Attributes["style"];
+		var hasImportantInOriginalStyle = originalStyleAttr != null && originalStyleAttr.Value.Contains("!important");
 
-			AddSpecialPremailerAttributes(attributeCssList, styleClass);
+		AddSpecialPremailerAttributes(attributeCssList, styleClass);
 
-			if (styleClass.Attributes.Count > 0)
-				attributeCssList.Add(new AttributeToCss { AttributeName = "style", CssValue = styleClass.ToString() });
+		if (styleClass.Attributes.Count > 0)
+			attributeCssList.Add(new AttributeToCss { AttributeName = "style", CssValue = styleClass.ToString(emitImportant: true) });
 
-			attributeCssList.AddRange(CssStyleEquivalence.FindEquivalent(domElement, styleClass));
+		attributeCssList.AddRange(CssStyleEquivalence.FindEquivalent(domElement, styleClass));
 
-			return attributeCssList;
+		if (hasImportantInOriginalStyle)
+		{
+			var styleAttr = attributeCssList.FirstOrDefault(a => a.AttributeName == "style");
+			if (styleAttr != null && !styleAttr.CssValue.Contains("!important"))
+			{
+				var parser = new CssParser();
+				var originalStyleClass = parser.ParseStyleClass("inline", originalStyleAttr.Value);
+				
+				var currentStyleClass = parser.ParseStyleClass("inline", styleAttr.CssValue);
+				
+				foreach (var attr in originalStyleClass.Attributes)
+				{
+					if (attr.Important && !currentStyleClass.Attributes.ContainsKey(attr.Style))
+					{
+						currentStyleClass.Attributes.Merge(attr);
+					}
+				}
+				
+				styleAttr.CssValue = currentStyleClass.ToString(emitImportant: true);
+			}
 		}
 
+		return attributeCssList;
+	}
+
```

Comment:
Please have a look and simplify this logic where possible. Think step by step, test-driven, write a unit test first in the `CssElementStyleResolverTests` file and make sure it covers the entire API here.